django/contrib/auth/hashers.py

```diff
@@ -416,35 +447,52 @@ def encode(self, password, salt):
         data = bcrypt.hashpw(password, salt)
         return "%s$%s" % (self.algorithm, data.decode('ascii'))
 
-    def verify(self, password, encoded):
-        algorithm, data = encoded.split('$', 1)
+    def decode(self, encoded):
+        algorithm, empty, algostr, work_factor, data = encoded.split('$', 4)
         assert algorithm == self.algorithm
+        salt, checksum = data[:22], data[22:]
+        return {
+            'algorithm': algorithm,
+            'algostr': algostr,
+            'work_factor': work_factor,
+            'salt': salt,
+            'checksum': checksum,
+        }
+
+    def verify(self, password, encoded):
+        decoded_entry = self.decode(encoded)
+        assert decoded_entry['algorithm'] == self.algorithm
+        partial = "".join((decoded_entry['salt'],
+                           decoded_entry['checksum']))
+        data = "$".join(('', decoded_entry['algostr'],
+                        decoded_entry['work_factor'],
+                        partial))
         encoded_2 = self.encode(password, data.encode('ascii'))
         return constant_time_compare(encoded, encoded_2)
 
     def safe_summary(self, encoded):
-        algorithm, empty, algostr, work_factor, data = encoded.split('$', 4)
-        assert algorithm == self.algorithm
-        salt, checksum = data[:22], data[22:]
+        decoded_entry = self.decode(encoded)
+        assert decoded_entry['algorithm'] == self.algorithm
         return {
-            _('algorithm'): algorithm,
-            _('work factor'): work_factor,
-            _('salt'): mask_hash(salt),
-            _('checksum'): mask_hash(checksum),
+            _('algorithm'): decoded_entry['algorithm'],
+            _('work_factor'): decoded_entry['work_factor'],
```

Comment:
Sorry, I had a typo in https://github.com/django/django/pull/12675#discussion_r418965162 -- what I ment to say was: keep `work factor` in safe summary (since it's for display and already translated) and `work_factor` in `decode`. This will also require less changes in the tests.

Reply:
>v would be the version of argon2i which is present in newer versions of the library. Where do you have problems handling this?

Sorry. Come again? I didn't have issues with the argon decode function.

> Sorry, I had a typo in #12675 (comment) -- what I ment to say was: keep work factor in safe summary (since it's for display and already translated) and work_factor in decode. This will also require less changes in the tests.

All good. Pushed a commit to change the string in safe summary.